android/gradle/wrapper/gradle-wrapper.properties

```diff
@@ -2,6 +2,6 @@
 #   $ ./gradlew wrapper --distribution-type=all --gradle-version=NEW_VERSION
 distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-5.5.1-all.zip
+distributionUrl=https\://services.gradle.org/distributions/gradle-6.0.1-all.zip
```

Comment:
> In general it's best to upgrade straight to the latest patch/bugfix release in a given series

Mmm, true. In this case, I took an approach of following the React Native changes commit-for-commit (as we do with everything else, to preserve detail or even add more detail), and also of checking the upgrades in those commits against our own procedure for upgrading Gradle, which is to run the `./gradlew` command shown at the top of `gradle-wrapper.properties`. This might reveal whether our upgrade procedure misaligns with theirs in some systematic way (the procedures seem to align well, at least in these upgrades), and also highlight any deviations RN seems to make from their procedure.

This approach confirms one thing that's already obvious: the comments added in android/gradle.properties from react-native@be2a2529a (`# AndroidX package structure [...]`) were definitely added by a human, not by running an upgrade command.

Things got a little weird when I realized that I couldn't just run our command to go from v5.5.1 to v6.0.1 and get the same result as I would following all the incremental upgrades (modulo the human-added comment I noted above). I suspect this is something common to our upgrade procedure and the one React Native uses.

In particular, take a look at this diff in `android/gradlew` from facebook/react-native@b1c954b1f (going from v5.6 to v5.6.2):

```diff
-# For Cygwin, switch paths to Windows format before running java
-if $cygwin ; then
+# For Cygwin or MSYS, switch paths to Windows format before running java
+if [ "$cygwin" = "true" -o "$msys" = "true" ] ; then
```

That change does get made if I do a v5.6.1 -> v5.6.2 upgrade. But:

- This automatic change does not get made if I do an all-at-once v5.5.1 -> v6.0.1 upgrade with our command.
- It's also left out if I do a v5.5.1 -> v5.6.2 upgrade.
- It's *also* left out if I do the usually reasonable v5.5.1 -> v5.6.4 upgrade. (Er...trying this again, it seems like that diff *does* get applied when I do the other leg, v5.6.4 -> v6.0.1...? Hmm.)
- Er, and now I look closer, it's actually left out when I do v5.6 -> v5.6.2—I realize this is the exact upgrade as facebook/react-native@b1c954b1f does...hmm. But what they probably did (that PR superseded facebook/react-native#26227), and what I probably did as well, is first do v5.6 -> v5.6.1, then, *without clearing that result*, do v5.6.1 -> v5.6.2.

I think as soon as I had doubts about a more obviously sensible strategy (all-at-once to v6.0.1 was what I'd thought of at first) complicating the path to a clearly correct reproduction of the RN changes, I decided to just follow all the incremental upgrades that they did. I also wanted to fill in some detail about potentially interesting changes with each one—but looking back, a lot of what I came up with is pretty boring, especially for the patch-version upgrades.

Reply:
> In particular, take a look at this diff

Hmm, that is some wacky behavior! I don't like that the upgrade seems to be path-dependent like that.

On staring at that a bit, I have a hypothesis:
* Each of the upgrades you tried where that diff didn't get applied were upgrades where you *started* from a version before v5.6.1; the one upgrade where it did was where you started from that version.
* Maybe the update to the `gradlew` script depends on the version you *start* from, rather than the one you upgrade to?

It's at least pretty easy to imagine how that behavior could arise -- you have one version installed, you ask that version to upgrade, but while that existing version is still running it also goes and tries to update your wrapper scripts to the latest and best that it knows about. Not *good* behavior but easy to imagine.

On that theory I'd expect `android/gradlew.bat` to behave the same way, for what that's worth.

---

I also took a look at the Gradle source code. It looks like this change was made first in https://github.com/gradle/gradle/commit/4b849cad3, for a source version of the wrapper script. Then in https://github.com/gradle/gradle/commit/89bce8f4ce0d8bf108dcbf3542d62167739f29e2 they updated the Gradle version used to build Gradle itself (one of many routine such updates), and the same commit also updated the wrapper scripts used for building Gradle itself, picking up the same change. Both of those happened before v5.6.0, so that's still a bit puzzling, though.

Reply:
> but looking back, a lot of what I came up with is pretty boring, especially for the patch-version upgrades, hmm.

Yeah. I think that's at the core of why I'd prefer to condense these -- there's a lot of detail here, and it makes it harder to focus on the most important information.

I think one structure I'd like for a commit message here would look something like this:
* This corresponds to the following five RN commits: [list the five commits]
* We treat this as an upgrade from 5.5.x to the latest 5.6.x and then to 6.0.1, which is the latest 6.0.x.
  * One of the RN upstream commits also added some comments in gradle.properties, which we reproduce.
  * We ran into a couple of different build failures; so we also add those to the troubleshooting doc, with the cache-clearing commands we used to resolve them (from SO: [link]).
* Release notes and upgrade guides, for 5.6.4 and 6.0.1.
  * Didn't spot any deprecations or breaking changes as applying to us; though hard to tell, and hope we'd catch any from just trying the build.
  * They recommend this build scan. First one found no deprecations, but second found those two things.
  * Also spotted at the console some deprecation warnings from the Android Gradle Plugin rather than Gradle itself, applying to these three of our dependencies (with our issues etc.).

One thing I like about that structure is that it collects all the information of a given type in one place, so it's easier to see the full list: in particular the list of upstream commits, and then the list of what was done in the commit. Relatedly, it puts those two things up front, and only after them the longer text describing things we looked at that (were important to check but) *didn't* end up resulting in any change appearing in this commit.

I expect it'd also turn out somewhat shorter; mainly by suppressing details about the intermediate patch versions, but partly also because when all the information of a given type is together that saves some repetition in labeling and framing it.

Reply:
Sure, sounds good! And thanks for the investigation!